### PR TITLE
Don't fail DagRun when leaf mapped_task is SKIPPED

### LINKED ISSUE #25200: DAG Run fails when chaining multiple empty mapped tasks
### Apache Airflow version

2.3.3 (latest released)

### What happened

On Kubernetes Executor and Local Executor (others not tested) a significant fraction of the DAG Runs of a DAG that has two consecutive mapped tasks which are are being passed an empty list are marked as failed when all tasks are either succeeding or being skipped.

image

### What you think should happen instead

The DAG Run should be marked success.

### How to reproduce

Run the following DAG on Kubernetes Executor or Local Executor.

The real world version of this DAG has several mapped tasks that all point to the same list, and that list is frequently empty. I have made a minimal reproducible example.

```py
from datetime import datetime

from airflow import DAG
from airflow.decorators import task


with DAG(dag_id="break_mapping", start_date=datetime(2022, 3, 4)) as dag:

    @task
    def add_one(x: int):
        return x + 1

    @task
    def say_hi():
        print("Hi")


    added_values = add_one.expand(x=[])
    added_more_values = add_one.expand(x=[])
    say_hi() >> added_values
    added_values >> added_more_values
```

### Operating System

Debian Bullseye

### Versions of Apache Airflow Providers

```
apache-airflow-providers-amazon==1!4.0.0
apache-airflow-providers-cncf-kubernetes==1!4.1.0
apache-airflow-providers-elasticsearch==1!4.0.0
apache-airflow-providers-ftp==1!3.0.0
apache-airflow-providers-google==1!8.1.0
apache-airflow-providers-http==1!3.0.0
apache-airflow-providers-imap==1!3.0.0
apache-airflow-providers-microsoft-azure==1!4.0.0
apache-airflow-providers-mysql==1!3.0.0
apache-airflow-providers-postgres==1!5.0.0
apache-airflow-providers-redis==1!3.0.0
apache-airflow-providers-slack==1!5.0.0
apache-airflow-providers-sqlite==1!3.0.0
apache-airflow-providers-ssh==1!3.0.0
```

### Deployment

Astronomer

### Deployment details

Local was tested on docker compose (from astro-cli)

### Anything else

_No response_

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by ashb**
This one was a fun one to track down, and was only a problem when the scheduler "expanded" the first mapped task (making it SKIPPED).

- The scheduler looks at `add_one` and marks it as SKIPPED.
- `unfinished_tis` contains `add_one__1`, and `_are_premature_tis` changes the state of `add_one__1`, (which just so happens to be a leaf task), and the check on line 584 essentially gets confused, as no one envisaged the states changing!

In a reverse of how this normally plays out, if the mini_scheduler in the LocalTaskJob was disabled then this example DAG would deadlock every time. (Since that mini scheduler only operates on a partial DAG it can't ever change the whole DagRun state.)

Fixes #25200

Ephraim had the idea of the fix, I just optomized it to not check the (possibly) large list everytime but only when a state was changed.

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

**Review by ephraimbuddy — APPROVED** [review]

**Comment by ashb** [comment]
Mypy is confusing me:

```
airflow/ti_deps/deps/trigger_rule_dep.py:149: error: Cannot determine type of
"trigger_rule"  [has-type]
            trigger_rule = cast(TR, task.trigger_rule)
```

File: airflow/models/baseoperator.py [hunk 1]
```diff
@@ -835,7 +835,7 @@ def __init__(
                 f"'{dag.dag_id if dag else ''}.{task_id}'; received '{trigger_rule}'."
             )
 
-        self.trigger_rule = TriggerRule(trigger_rule)
+        self.trigger_rule: TriggerRule = TriggerRule(trigger_rule)
```
Comment (ashb) [inline airflow/models/baseoperator.py:838]: Mypy really can't work this out? :o

Comment (uranusjr) [reply airflow/models/baseoperator.py:838]: I think it has something to do with either (or a combination of) Enum having a custom metaclass, TriggerRule does multi-inheritance, and one of the parents is a built-in.

**Pull request merged**